File: android/src/main/java/tpf/expo/veriff/ExpoVeriffModule.kt
package tpf.expo.veriff

import com.veriff.GeneralConfig
import com.veriff.Result
import android.app.Activity
import com.veriff.Configuration
import com.veriff.Sdk.createLaunchIntent
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.Locale

class ExpoVeriffModule : Module() {
  private var callbackResult: Promise? = null
  private var isSessionActive = false

  fun launchVeriff(
    sessionUrl: String,
    p: Promise
  ) {
    if (callbackResult != null || isSessionActive) {
      p.reject("ALREADY_CREATED", "Veriff session is already active", null)
      return
    }

    val activity: Activity? = appContext.currentActivity
    if (activity == null) {
      p.reject("ACTIVITY_NOT_FOUND", "Activity not found", null)
      return
    }

    try {
      val appLocale = Locale.ENGLISH
      val configuration = Configuration.Builder()
              .locale(appLocale)
              .build()
      val intent = createLaunchIntent(activity, sessionUrl, configuration)
      activity.startActivityForResult(intent, REQUEST_CODE)
      callbackResult = p
      isSessionActive = true
    } catch (e: Exception) {
      p.reject("SETUP_ERROR", "Failed to start Veriff session: ${e.message}", null)
    }
  }

  private fun cleanup() {
    callbackResult = null
    isSessionActive = false
  }

  override fun definition() = ModuleDefinition {
    Name("ExpoVeriff")

    AsyncFunction("launchVeriff") { sessionUrl: String, promise: Promise ->
      launchVeriff(sessionUrl, promise)
    }

    OnActivityResult { _, payload ->
      if (payload.requestCode == REQUEST_CODE) {
        try {
          val veriffResult = Result.fromResultIntent(payload.data)

          when (veriffResult?.status) {
            Result.Status.DONE -> {
              val token = payload.data?.getStringExtra(GeneralConfig.INTENT_EXTRA_SESSION_URL)
              callbackResult?.resolve(token)
            }
            Result.Status.CANCELED -> {
              callbackResult?.reject("CANCELED", "User canceled the verification", null)
            }
            Result.Status.ERROR -> {
              // Generic error handling without relying on specific enum values
              val errorCode = veriffResult.error?.name ?: "UNKNOWN_ERROR"
              val errorMessage = when (errorCode) {
                "CAMERA_UNAVAILABLE" -> "Camera is not available or permission denied"
                "MICROPHONE_UNAVAILABLE" -> "Microphone is not available or permission denied"
                "NETWORK_ERROR" -> "Network connection error occurred"
                "SERVER_ERROR" -> "Server error occurred"
                "UPLOAD_ERROR" -> "Upload error occurred"
                "VIDEO_FAILED" -> "Video processing failed"
                "LOCAL_ERROR" -> "Local error occurred"
                "DEPRECATED_SDK_VERSION" -> "Veriff SDK version is not supported"
                "UNKNOWN" -> "An unknown error occurred during verification"
                else -> "An unknown error occurred during verification"
              }
              callbackResult?.reject(errorCode, errorMessage, null)
            }
            null -> {
              callbackResult?.reject("UNKNOWN_ERROR", "Failed to process Veriff result", null)
            }
          }
        } catch (e: Exception) {
          callbackResult?.reject("UNKNOWN_ERROR", "Failed to process Veriff result: ${e.message}", null)
        } finally {
          cleanup()
        }
      }
    }
  }

  companion object {
    private const val REQUEST_CODE = 47239
  }
}

File: ios/ExpoVeriffModule.swift
import ExpoModulesCore
import Veriff
import AVFoundation

public class ExpoVeriffModule: Module {
    private var promise: Promise?
    private var isSessionActive = false
    
    func launchVeriff(
        sessionUrl: String,
        p: Promise
    ) {
        if (self.promise != nil || isSessionActive) {
            p.reject("ALREADY_CREATED", "Veriff session is already active")
            return
        }
        
        // Set flags IMMEDIATELY to prevent race condition
        self.isSessionActive = true
        self.promise = p
        
        // Check camera permissions first
        checkCameraPermissions { [weak self] hasPermission in
            guard let self = self else { return }
            
            if !hasPermission {
                DispatchQueue.main.async {
                    p.reject("CAMERA_PERMISSION_DENIED", "Camera permission is required for Veriff")
                    self.cleanup()
                }
                return
            }
            
            DispatchQueue.main.async {
                let locale = Locale(identifier: "en")
                let configuration = Veriff.VeriffSdk.Configuration(languageLocale: locale)
                let veriff = Veriff.VeriffSdk.shared
                veriff.delegate = self
                veriff.implementationType = .reactNative
                
                do {
                    veriff.startAuthentication(sessionUrl: sessionUrl, configuration: configuration)
                } catch {
                    self.promise?.reject("SETUP_ERROR", "Failed to start Veriff session: \(error.localizedDescription)")
                    self.cleanup()
                }
            }
        }
    }
    
    private func checkCameraPermissions(completion: @escaping (Bool) -> Void) {
        switch AVCaptureDevice.authorizationStatus(for: .video) {
        case .authorized:
            completion(true)
        case .notDetermined:
            AVCaptureDevice.requestAccess(for: .video) { granted in
                DispatchQueue.main.async {
                    completion(granted)
                }
            }
        case .denied, .restricted:
            completion(false)
        @unknown default:
            completion(false)
        }
    }
    
    private func cleanup() {
        promise = nil
        isSessionActive = false
    }
    
  public func definition() -> ModuleDefinition {
    Name("ExpoVeriff")

    AsyncFunction("launchVeriff") { (sessionUrl: String, p: Promise) in
        self.launchVeriff(sessionUrl: sessionUrl, p: p)
    }
  }
}

extension ExpoVeriffModule: VeriffSdkDelegate {
    public func sessionDidEndWithResult(_ result: Veriff.VeriffSdk.Result) {
        defer { cleanup() }
        
        switch result.status {
        case .done:
            self.promise?.resolve(result.sessionUrl.absoluteURL)
        case .canceled:
            self.promise?.reject("CANCELED", "User canceled the verification")
        case .error(let err):
            switch err {
                case .cameraUnavailable:
                    self.promise?.reject("CAMERA_UNAVAILABLE", "Camera is not available or permission denied")
                case .microphoneUnavailable:
                    self.promise?.reject("MICROPHONE_UNAVAILABLE", "Microphone is not available or permission denied")
                case .networkError,
                     .uploadError:
                    self.promise?.reject("NETWORK_ERROR", "Network connection error occurred")
                case .serverError,
                     .videoFailed,
                     .localError:
                    self.promise?.reject("SETUP_ERROR", "Failed to setup or process verification")
                case .deprecatedSDKVersion:
                    self.promise?.reject("UNSUPPORTED_SDK_VERSION", "Veriff SDK version is not supported")
                case .unknown:
                    self.promise?.reject("UNKNOWN_ERROR", "An unknown error occurred during verification")
                @unknown default:
                    self.promise?.reject("UNKNOWN_ERROR", "An unknown error occurred during verification")
            }
        @unknown default:
            self.promise?.reject("UNKNOWN_ERROR", "An unknown error occurred during verification")
        }
    }
}